tasks/libs/common/remote_api.py

```diff
@@ -0,0 +1,102 @@
+
+        By default, the request method is GET, or POST if data is not empty.
+        method: Can be set to GET, POST, PUT or DELETE to force the REST method used.
+
+        By default, we return the text field of the response object. The following fields can alter this behavior:
+        json_output: the json field of the response object is returned.
+        stream_output: the request asks for a stream response, and the raw response object is returned.
+        raw_output: the content field of the resposne object is returned.
+        """
+        import requests
+
+        url = self.BASE_URL + path
+
+        # TODO: Use the param argument of requests instead of handling URL params
+        # manually
+        try:
+            # If json_input is true, we specifically want to send data using the json
+            # parameter of requests.post / requests.put
+            if method == "PUT":
+                if json_input:
+                    r = requests.put(url, headers=headers, json=data, stream=stream_output)
+                else:
+                    r = requests.put(url, headers=headers, data=data, stream=stream_output)
+            elif method == "DELETE":
+                r = requests.delete(url, headers=headers, stream=stream_output)
+            elif data or method == "POST":
+                if json_input:
+                    r = requests.post(url, headers=headers, json=data, stream=stream_output)
+                else:
+                    r = requests.post(url, headers=headers, data=data, stream=stream_output)
+            else:
+                r = requests.get(url, headers=headers, stream=stream_output)
+            if r.status_code == 401:
+                print(self.authorization_error_message)
+
+                print("{} says: {}".format(self.api_name, r.json()["error_description"]))
+                raise Exit(code=1)
+        except requests.exceptions.Timeout:
+            print("Connection to {} ({}) timed out.".format(self.api_name, url))
+            raise Exit(code=1)
+        except requests.exceptions.RequestException as e:
+            m = errno_regex.match(str(e))
+            if not m:
+                print("Unknown error raised connecting to {} ({}): {}".format(self.api_name, url, e))
+
+            # Parse errno to give a better explanation
+            # Requests doesn't have granularity at the level we want:
+            # http://docs.python-requests.org/en/master/_modules/requests/exceptions/
+            errno_code = int(m.group(1))
+            message = m.group(2)
+
+            if errno_code == errno.ENOEXEC:
+                print("Error resolving {}: {}".format(url, message))
+            elif errno_code == errno.ECONNREFUSED:
+                print("Connection to {} ({}) refused".format(self.api_name, url))
+            else:
+                print("Error while connecting to {}: {}".format(url, str(e)))
+            raise Exit(code=1)
+
+        if json_output:
+            return r.json()
```

Comment:
What do you think about using this https://stackoverflow.com/a/15882054 to return a class instead of a dictionary? I don't feel like this would strongly improve things but it feels like a step in the right direction IMO (ideally what would happen is we have concrete types for everything but that would be too much effort, this can be a starting point)

Reply:
I agree it'd be a good thing, however we may want to do it in a separate PR, as this would add a bunch of new changes to an already big PR. I added a card to tackle that (AP-1278).